### PR TITLE
Implement Experimental GraalJsEngine

### PR DESCRIPTION
**Pull request by Leland-Takamine**
# Proposed Changes

Enables opting in to use GraalJS instead of Rhino JS for javascript evaluation. GraalJS is fully ECMAScript 2022 compliant while Rhino JS only supports ES5.

This feature is experimental for now, but we intend to make GraalJS the default if all goes well.

### Enable GraalJS via Flow config

> **Note**
> This only has an effect if present at the top-level flow. It is ignored when included in any subflows.

```yaml
appId: com.example.app
jsEngine: graaljs
---
# Note: The ?? operator is an example of an ES2020 feature and is not supported by Rhino JS
- inputText: ${null ?? 'foo'}
```

### Enable GraalJS via environment variable

> **Warning**
> This env var will have no effect when running on Maestro Cloud. Use the Flow config above instead to opt into GraalJS on Maestro Cloud.

```shell
export MAESTRO_USE_GRAALJS=true
maestro test my-flow.yaml
```

# GraalJS behavior differences

There are some differences between the new `GraalJsEngine` and the current `RhinoJsEngine` implementation that are worth noting. All of the differences below and some others are documented and tested by the `GraalJsEngineTest` and `RhinoJsEngineTest` tests.

**tldr; The variable scoping with the GraalJS implementation is more consistent and understandable**

<table>
<tr><th>Rhino JS</th><th>GraalJS</th></tr>
<tr>
<td>
<li><b>Can not</b> reuse variable names across scripts within the same Flow</li>
<li><b>Can</b> reuse variable names if redeclaration lives in a subflow</li>
<li><b>Can</b> access variables declared earlier in the Flow in a separate script</li>
<li><code>output</code> variable is shared across all scripts</li>
</td>
<td>
<li>Variables are local to each script</li>
<li><code>output</code> variable is shared across all scripts</li>
</td>
</tr>
</table>

**Some examples**

<table>
<tr><th></th><th>Example</th><th>Rhino JS</th><th>GraalJS</th></tr>

<tr>
  <td><b>Redeclaring variables across scripts</b></td>
  <td>
  <pre>
appId: com.example
---
- evalScript: ${const foo = null}
- evalScript: ${const foo = null}
</pre>
  </td>
  <td>❌ Variable redeclarations throw an error</td>
  <td>✅ Variable names can be reused across scripts</td>
</tr>

<tr>
  <td><b>Accessing variables across scripts</b></td>
  <td>
  <pre>
appId: com.example
---
- evalScript: ${const name = 'joe'}
- assertTrue: ${name === 'joe'}
</pre>
  </td>
  <td>✅ Variables are accessible across scripts</td>
  <td>❌ Variables can't be accessed across scripts</td>
</tr>

<tr>
  <td><b>Special character handling</b></td>
  <td>
  <pre>
appId: com.example.app
env:
  FOO: \
---
- inputText: ${FOO}
</pre>
  </td>
  <td>❌ Single backslash causes an exception</td>
  <td>✅ Single backslash and all other special chars are handled correctly</td>
</tr>

</table>

# Testing

Ran integration tests with `MAESTRO_USE_GRAALJS=true`. 2 tests failed due to known differences between GraalJS and RhinoJS described above. All other tests passed.

### CODE REVIEW

File: maestro-orchestra/src/main/java/maestro/orchestra/Orchestra.kt [hunk 1]
```diff
@@ -165,6 +168,15 @@ class Orchestra(
         return true
     }
 
+    @Synchronized
+    private fun initJsEngine(config: MaestroConfig?) {
+        if (this::jsEngine.isInitialized) {
+            jsEngine.close()
+        }
+        val shouldUseGraalJs = config?.ext?.get("jsEngine") == "graaljs" || System.getenv("MAESTRO_USE_GRAALJS") == "true"
+        jsEngine = if (shouldUseGraalJs) GraalJsEngine() else RhinoJsEngine()
```
Comment (Leland-Takamine) [inline maestro-orchestra/src/main/java/maestro/orchestra/Orchestra.kt:177]: Note, we are now creating a new instance of the JsEngine on every flow execution. Before we were caching the JsEngine instance. This shouldn't have any significant impact on performance.

**Comment by Leland-Takamine** [comment]
Addresses #1081

File: maestro-client/src/main/java/maestro/js/GraalJsEngine.kt [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        env: Map<String, String>,
+        sourceName: String,
+        runInSubScope: Boolean,
+    ): Value {
+        envBinding.putAll(env)
+        val source = Source.newBuilder("js", script, sourceName).build()
+        return createContext().eval(source)
+    }
+
+    private fun createContext(): Context {
```
Comment (Leland-Takamine) [inline maestro-client/src/main/java/maestro/js/GraalJsEngine.kt:70]: Creates a new context for each script evaluation but this is a fast operation based on my adhoc testing

File: maestro-client/src/main/java/maestro/js/GraalJsEngine.kt [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+            .logHandler(NULL_HANDLER)
+            .out(outputStream)
+            .build()
+
+        openContexts.add(context)
+
+        envBinding.forEach { (key, value) -> context.getBindings("js").putMember(key, value) }
+
+        context.getBindings("js").putMember("http", httpBinding)
+        context.getBindings("js").putMember("output", ProxyObject.fromMap(outputBinding))
```
Comment (Leland-Takamine) [inline maestro-client/src/main/java/maestro/js/GraalJsEngine.kt:91]: `output` and `maestro` are the two variables (aka "bindings") shared across scripts

File: maestro-test/src/test/kotlin/maestro/test/GraalJsEngineTest.kt [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+import com.google.common.truth.Truth
+import com.google.common.truth.Truth.assertThat
+import maestro.js.GraalJsEngine
+import org.graalvm.polyglot.PolyglotException
+import org.graalvm.polyglot.Value
+import org.junit.jupiter.api.BeforeEach
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.assertThrows
+
+class GraalJsEngineTest : JsEngineTest() {
```
Comment (Leland-Takamine) [inline maestro-test/src/test/kotlin/maestro/test/GraalJsEngineTest.kt:12]: These tests document known behaviors that are different from Rhino JS. Common tests live in the base JsEngineTest class.

File: maestro-test/src/test/kotlin/maestro/test/RhinoJsEngineTest.kt [hunk 5]
```diff
@@ -0,0 +1,58 @@
+package maestro.test
+
+import com.google.common.truth.Truth.assertThat
+import maestro.js.RhinoJsEngine
+import org.junit.jupiter.api.BeforeEach
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.assertThrows
+import org.mozilla.javascript.RhinoException
+
+class RhinoJsEngineTest : JsEngineTest() {
```
Comment (Leland-Takamine) [inline maestro-test/src/test/kotlin/maestro/test/RhinoJsEngineTest.kt:10]: These tests document known behaviors that are different from GraalJS. Common tests live in the base JsEngineTest class.

**Review by axelniklasson — APPROVED** [review]
Nice, works like a charm!

**Review by igorsmotto — APPROVED** [review]

**Comment by igorsmotto** [comment]
Should the env var `MAESTRO_USE_GRAALJS` even exists? Given that it only works on Maestro (not on Cloud) and there is already a place to set the JsEngine (MaestroConfig)?

**Comment by Leland-Takamine** [comment]
Hm I was was thinking the env var would be useful for testing an entire test suite with graaljs without editing the flow file. This may come in handy when testing against existing test suites (which I think is useful for both Maestro users and for us). I agree that it has the potential to cause confusion but on the other hand it will only exist while graaljs is experimental. But I'm not married to it so we can remove if it seems too problematic

**Comment by Leland-Takamine** [comment]
@igorsmotto I was thinking about that but I wanted to also support this case: `maestro test myflow.yaml` which doesn't use a config file

**Comment by Leland-Takamine** [comment]
@igorsmotto I couldn't find a better alternative that doesn't require a lot more rearchitecting, and given that this is a temporary config while graaljs is still experimental, I've landed on the following compromise:

* Both the env var and flow are available to enable graaljs
* The env var only works locally
* The flow config only has an effect if present on the top-level flow
* **The flow config does not have an effect if present in a subflow**
